models/rnnt.py

```diff
@@ -0,0 +1,202 @@
+from tinygrad.tensor import Tensor
+from tinygrad.nn import Linear
+import numpy as np
+from extra.utils import download_file
+from pathlib import Path
+
+
+class RNNT:
+  def __init__(self, input_features=240, vocab_size=29, enc_hidden_size=1024, pred_hidden_size=320, joint_hidden_size=512, pre_enc_layers=2, post_enc_layers=3, pred_layers=2, stack_time_factor=2, dropout=0.32):
+    self.encoder = Encoder(input_features, enc_hidden_size, pre_enc_layers, post_enc_layers, stack_time_factor, dropout)
+    self.prediction = Prediction(vocab_size, pred_hidden_size, pred_layers, dropout)
+    self.joint = Joint(vocab_size, pred_hidden_size, enc_hidden_size, joint_hidden_size, dropout)
+
+  def __call__(self, x, x_lens, y, h=None, c=None):
+    f, _ = self.encoder(x, x_lens)
+    g, _, _ = self.prediction(y, h, c)
+    out = self.joint(f, g)
+    return out
+
+  def decode(self, x, x_lens):
+    logits, logit_lens = self.encoder(x, x_lens)
+    outputs = []
+    for b in range(logits.shape[0]):
+      inseq = logits[b, :, :].unsqueeze(1)
+      logit_len = logit_lens[b]
+      seq = self._greedy_decode(inseq, logit_len)
+      outputs.append(seq)
+    return outputs
+
+  def _greedy_decode(self, logits, logit_lens):
+    h = c = None
+    labels = []
+    for time_idx in range(int(logit_lens.numpy().item())):
+      logit = logits[time_idx, :, :].unsqueeze(0)
+      not_blank = True
+      added = 0
+      while not_blank and added < 30:
+        label = Tensor([[labels[-1] if labels[-1] <= 28 else labels[-1] - 1]]) if len(labels) > 0 else None
+        g, h_, c_ = self.prediction(label, h, c)
+        j = self.joint(logit, g)[:, 0, 0, :][0, :].numpy()
+
+        k = np.argmax(j, axis=0)
+        not_blank = k != 28
+        if not_blank:
+          labels.append(k)
+          h, c = h_, c_
+        added += 1
+    return labels
+
+  def load_from_pretrained(self):
+    fn = Path(__file__).parent.parent / "weights/rnnt.pt"
+    download_file("https://zenodo.org/record/3662521/files/DistributedDataParallel_1576581068.9962234-epoch-100.pt?download=1", fn)
+
+    import torch
+    state_dict = torch.load(open(fn, "rb"), map_location="cpu")["state_dict"]
+
+    # encoder
```

Comment:
See how the other models do this with get_child. Any reason this one can't be done like that? The code is a lot shorter

Reply:
torch uses `weights_ih_l{layer_number}` for their lstm layer weights which can't really be done cleanly